lalrpop/src/build/mod.rs

```diff
@@ -205,42 +207,29 @@ fn needs_rebuild(lalrpop_file: &Path, rs_file: &Path) -> io::Result<bool> {
 
 fn lalrpop_files<P: AsRef<Path>>(root_dir: P) -> io::Result<Vec<PathBuf>> {
     let mut result = vec![];
-    for entry in fs::read_dir(root_dir)? {
+
+    let walkdir = WalkDir::new(root_dir)
+        .follow_links(true)
+        // Use deterministic ordering:
+        .sort_by_file_name()
+        // Prevent loops:
+        .same_file_system(true);
+    for entry in walkdir {
         let entry = entry?;
-        let file_type = entry.file_type()?;
+        // `file_type` follows symlinks, so if `entry` points to a symlink to a file, then
+        // `is_file` returns true.
+        if !entry.file_type().is_file() {
+            continue;
+        }
 
         let path = entry.path();
-
-        if path.is_dir() {
-            result.extend(lalrpop_files(&path)?);
+        if path.extension() != Some(OsStr::new("lalrpop")) {
+            continue;
         }
 
-        let is_valid_symlink_file = || -> bool {
-            if !file_type.is_symlink() {
-                false
-            } else {
-                // Ensure all symlinks are resolved to a file
-                // Or ignore erroneous ones https://github.com/lalrpop/lalrpop/issues/808
```

Comment:
This implementation will probably regress on this issue?

If it's easy to skip over bad links then maybe we should keep that behavior. At the very least, I would like to make sure that if lalrpop does error out on a bad link, it contains the name.

Reply:
You're right, this was a regression. I've added some handling code that should restore the current behavior: We print a warning about the dangling symlink and otherwise ignore it.

Reply:
Your solution here looks reasonable to me. Just one last think about symlink cycles.